include/hex.hpp

```diff
@@ -0,0 +1,169 @@
+HexCoords hexRD (int x);
+HexCoords hexLD (int x);
+HexCoords hexL  (int x);
+
+
+// Edges of a hex
+enum class Edge {
+    RightUp = 0, RU = 0,
+    Right = 1, R = 1,
+    RightDown = 2, RD = 2,
+    LeftDown = 3, LD = 3,
+    Left = 4, L = 4,
+    LeftUp = 5, LU = 5
+};
+
+constexpr static float sqrt3 = 1.73205080757; // comes up in hex maths
+// because fucking modulo operator
+inline int positive_modulo(int i, int n) {
+    return (i % n + n) % n;
+}
+
+struct HexCoords {
+    int q;
+    int r;
+    int s;
+
+    bool operator== (const HexCoords& other) const;
+    static HexCoords from_axial(int q, int r);
+    HexCoords operator+ (const HexCoords& other) const;
+    HexCoords operator- (const HexCoords& other) const;
+    EdgeCoords operator+ (const Edge& edge) const;
+    std::array<HexCoords, 6> neighbours() const;
+    int distance (const HexCoords& to) const;
+    std::pair<float, float> to_world_unscaled () const;
+    static HexCoords rounded_to_hex(float q, float r, float s);
+    static HexCoords from_world_unscaled(float x, float y);
+    static std::vector<HexCoords> make_line(const HexCoords from, const HexCoords to);
+};
+
+struct EdgeCoords {
+    HexCoords hex;
+    Edge edge;
+};
+
+template <typename HexT>
+struct CylinderHexWorld {
+    int width;
+    int height;
+    HexT empty_hex;
+    std::vector<HexT> data;
+
+    CylinderHexWorld (int width, int height, HexT default_hex, HexT empty_hex)
+        : width(width), height(height), empty_hex(empty_hex)
+    {
+        data.resize((width)*(height), default_hex);
+    }
+
+    HexCoords normalized_coords (const HexCoords abnormal) const {
+        const auto q = positive_modulo(abnormal.q, width);
+        const auto r = std::max(std::min<int>(abnormal.r, height), 0);
+        const auto s = -q-r;
+        return HexCoords{q, r, s};
+    }
+
+    void normalize_height(HexCoords& hc) const {
+        hc.r = std::max(std::min<int>(hc.r, height), 0);
+    }
+
+    int compute_index (const HexCoords hc) const {
+        int direct_row = hc.r;
+        int direct_column = hc.q;
+        return direct_row*width + direct_column;
+    }
+
+    int compute_normalized_index (const HexCoords ahc) const {
+        const auto hc = normalized_coords(ahc);
+        return compute_index(hc);
+    }
+
+    HexT at(const HexCoords hc) {
+        return at_ref_abnormal(hc).value_or(empty_hex);
+    }
+
+    std::optional<std::reference_wrapper<HexT>> at_ref_abnormal(const HexCoords hc) {
+        if (hc.r < 0 || hc.r >= height) {
+            return {};
+        }
+        auto wrapped = hc;
+        wrapped.q = positive_modulo(wrapped.q, width);
+        return data.at(compute_index(wrapped));
+    }
+
+    HexT& at_ref_normalized(const HexCoords hc) {
+        return data.at(compute_index(normalized_coords(hc)));
+    }
+
+    std::vector<HexCoords> all_within_unscaled_quad 
+    (Vector2 top_left, Vector2 top_right, Vector2 bottom_left, Vector2 bottom_right) 
+    {
```

Comment:
formatowanie do poprawy, bo jest bardzo niejednolite, juź lepiej by było
    std::vector<HexCoords> all_within_unscaled_quad (
Vector2 top_left,
 Vector2 top_right,
 Vector2 bottom_left,
 Vector2 bottom_right
){

Jak klamerki otwieramy w tej samej linii to nie zaczynajmy tutaj inaczej